framework/src/kernels/NonlocalKernel.C

```diff
@@ -0,0 +1,102 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "NonlocalKernel.h"
+#include "Assembly.h"
+#include "MooseVariable.h"
+#include "Problem.h"
+#include "SubProblem.h"
+#include "SystemBase.h"
+
+// libmesh includes
+#include "libmesh/threads.h"
+#include "libmesh/quadrature.h"
+
+template<>
+InputParameters validParams<NonlocalKernel>()
+{
+  InputParameters params = validParams<Kernel>();
+  return params;
+}
+
+NonlocalKernel::NonlocalKernel(const InputParameters & parameters) :
+    Kernel(parameters)
+{
+}
+
+void
+NonlocalKernel::computeJacobian()
+{
+  Kernel::computeJacobian();
+
+  DenseMatrix<Number> & keg = _assembly.jacobianBlockNonlocal(_var.number(), _var.number());
+  _nonlocal_ke.resize(keg.m(), keg.n());
+  _nonlocal_ke.zero();
+
+  // compiling set of global IDs for the local DOFs on the element
+  std::set<dof_id_type> local_dofindices(_var.dofIndices().begin(), _var.dofIndices().end());
+  // storing the global IDs for all the DOFs of the variable
+  std::vector<dof_id_type> var_alldofindices(_var.allDofIndices());
```

Comment:
Should be:

`const std::vector<dof_id_type> & var_alldofindices = _var.allDofIndices();`